engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/UniqueItemsValidator.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.constraintvalidators.hv;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import org.hibernate.validator.constraints.UniqueItems;
+
+/**
+ * Validates that the provided collection is a set of unique items
+ * <p>
+ * Uniqueness is defined by the equals method of the item being compared.
+ *
+ * @author Tadhg Pearson
+ */
+public class UniqueItemsValidator implements ConstraintValidator<UniqueItems, Collection<?>> {
+	@Override
+	public void initialize(UniqueItems uniqueItems) {
+		//No annotation fields to initialize
+	}
+
+	/**
+	 * @param objects Potentially null collection of input objects to check for duplicates
+	 * @param validatorContext Non-null {@link ConstraintValidatorContext} to accumulate validation messages
+	 *
+	 * @return True if the input collection is null or does not contain duplicate objects
+	 */
+	@Override
+	public boolean isValid(Collection<?> objects, ConstraintValidatorContext validatorContext) {
+		boolean valid = isValid( objects );
+		if ( !valid ) {
+			List<String> duplicates = findDuplicates( objects );
+			String message = "Contains the following duplicate items: " + duplicates;
+			validatorContext.disableDefaultConstraintViolation();
+			validatorContext.buildConstraintViolationWithTemplate( message ).addConstraintViolation();
+		}
+		return valid;
+	}
+
+	private boolean isValid(Collection<?> objects) {
+		boolean valid;
+		if ( objects == null ) {
+			valid = true;
+		}
+		else {
+			Set<?> set = new HashSet<>( objects );
+			valid = set.size() == objects.size();
+		}
+		return valid;
+	}
+
+	/**
+	 * Find the duplicate items the provided input collection. The duplicate items will be presented in the order in which
+	 * they occur. Duplicate items which occur multiple times will be presented multiple times in the duplicates list.
+	 *
+	 * @param objects Non-null collection of objects in which to find duplicates.
+	 *
+	 * @return Non-null collection of the toString output for the dupliciate objects.
+	 */
+	private List<String> findDuplicates(Collection<?> objects) {
+		Set<Object> uniqueItems = new HashSet<>();
+		List<?> duplicates = objects.stream().filter( o -> !uniqueItems.add( o ) ).collect( Collectors.toList() );
+		return duplicates.stream().map( String::valueOf ).collect( Collectors.toList() );
```

Comment:
Is there a reason why you create an intermediate list here? You could continue to use the stream?

Reply:
Good feedback - there used to be a method in there - it got pulled back and the streams never got joined

Reply:
We discussed this with Marko this morning and we are wondering if, maybe, we shouldn't add the items to the message but rather provide the objects in a Payload (the payload API allows to add an object to the constraint violation that the user can get afterwards). That would allow the users to get the duplicate objects directly and do whatever they want with it instead of just having a string.

Calling `toString()` might be ugly if not implemented.

So, maybe, we could have a generic message saying that there are duplicate items and then the user could use the payload to do whatever he wants with it?

As the first user of this annotation, let us know what you think.

Btw, thinking a bit more about it, I think we should support having nulls in the collections, don't you think? That would qualify for a new test :).

Reply:
There are many options like that... allowing users to define their own comparators instead of equals seems like another obvious way to go. My philosophy here was start simple - rely on the existing equals and toString, and see if anyone care. It's easier to expand it

Reply:
Regarding null items in the collections, this is already supported. One null is unique, multiple nulls are not unique and cause an exception.

I can add this to the tests, should it also be documented?